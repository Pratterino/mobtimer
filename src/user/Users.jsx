import React, {Component, Fragment} from 'react';
import {bindActionCreators} from "redux";
import {connect} from "react-redux";
import {DragDropContext, Draggable, Droppable} from 'react-beautiful-dnd';
import {usersSelector} from "./userReducer";
import {addUser, nextUser, updateUserOrder} from "./userActions";
import Input from "./../Input";
import User from "./User";
import "./Users.scss";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faFastForward} from "@fortawesome/free-solid-svg-icons";

class Users extends Component {
    constructor(props) {
        super(props);
        this.state = {
            nameValue: "",
            users: this.props.users,
        };
        this.grid = 8;
    }

    componentWillReceiveProps(nextProps, nextContext) {
        if (nextProps.users !== this.state.users) {
            this.setState({users: nextProps.users});
        }
    }

    addUser = (event) => {
        event.preventDefault();
        this.props.addUser(event.target[0].value);
    };

    onDragEnd = (result) => {
        // dropped outside the list
        if (!result.destination) {
            return;
        }

        const users = this.reorder(
            this.state.users,
            result.source.index,
            result.destination.index
        );

        this.setState({
            users,
        });

        this.props.updateUserOrder(users);
    };

    getItemStyle = (isDragging, draggableStyle) => ({
        userSelect: 'none',
        background: isDragging ? getComputedStyle(document.body).getPropertyValue('--highlight-color') : 'transparent',
        // styles we need to apply on draggables
        ...draggableStyle,
    });

    // css
    getListStyle = isDraggingOver => ({
        background: isDraggingOver ? 'rgba(0,0,0,0.2)' : '',
        display: 'inline-flex',
        overflow: 'auto',
    });

    reorder = (list, startIndex, endIndex) => {
        const result = Array.from(list);
        const [removed] = result.splice(startIndex, 1);
        result.splice(endIndex, 0, removed);
        return result;
    };

    // end css

    render() {
        return (
            <Fragment>
                <DragDropContext onDragEnd={this.onDragEnd}>
                    <Droppable droppableId="droppable" direction="horizontal">
                        {(provided, snapshot) => (
                            <div
                                className="draggable-users"
                                ref={provided.innerRef}
                                style={this.getListStyle(snapshot.isDraggingOver)}
                                {...provided.droppableProps}
                            >
                                {this.state.users.map((user, index) => (
                                    <Draggable key={user.name} draggableId={user.name} index={index}>
                                        {(provided, snapshot) => (
                                            <div
                                                ref={provided.innerRef}
                                                {...provided.draggableProps}
                                                {...provided.dragHandleProps}
                                                style={this.getItemStyle(
                                                    snapshot.isDragging,
                                                    provided.draggableProps.style
                                                )}
                                            >
                                                <User
                                                    key={user.name}
                                                    user={user}
                                                    disabled={user.disabled}
                                                />
                                            </div>
                                        )}
                                    </Draggable>
                                ))}
                                {provided.placeholder}
                                <form className="user__add" onSubmit={(e) => e.preventDefault()}>
                                    <Input handleInputSubmit={(text) => this.props.addUser(text)}>Add user</Input>
                                </form>
                            </div>
                        )}
                    </Droppable>
                </DragDropContext>

                <section className="users__footer">
                    <FontAwesomeIcon
                        size={"2x"}
                        className="pointer"
                        onClick={this.props.nextUser}
                        icon={faFastForward}
                    />
                </section>
            </Fragment>
        );
    }
}

const mapStateToProps = state => ({
    timer: state.timer,
    users: usersSelector(state),
});
const mapDispatchToProps = dispatch => bindActionCreators({
    addUser,
    nextUser,
    updateUserOrder,
}, dispatch);


export default connect(mapStateToProps, mapDispatchToProps)(Users);
