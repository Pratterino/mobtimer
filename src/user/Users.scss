div[data-react-beautiful-dnd-drag-handle] {
  background: #61dafb;
  display: inline-flex;
}
