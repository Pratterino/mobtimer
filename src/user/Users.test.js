import React from 'react';
import {unwrapped as UnwrappedUsers} from './Users';
import {shallow} from 'enzyme';

describe("Users", () => {
    let wrapper;
    let addUserSpy;
    const users = [{
        name: "Bob Ducca",
    }, {
        name: "Dabney Coleperson",
    }, {
        name: "Llarn Bleb-Raag",
    }];

    const renderComponent = (props) => {
        addUserSpy = jest.fn();
        wrapper = shallow(
            <UnwrappedUsers
                users={users}
                addUser={addUserSpy}
                nextUser={jest.fn()}
                resetTimer={jest.fn()}
                updateUserOrder={jest.fn()}
                {...props}
            />
        );
    };

    beforeEach(() => {
        renderComponent();
    });

    describe('componentWillReceiveProps', () => {
        it('should not update local state', () => {
            let state = {
                users,
            };

            wrapper.instance().componentWillReceiveProps(state);
            expect(wrapper.state('users')).toEqual(users);
        });

        it('should update local state', () => {
            const newUsers = [...users, {name: "Jenny"}];

            let newState = {
                users: newUsers,
            };

            wrapper.instance().componentWillReceiveProps(newState);
            expect(wrapper.state('users')).toEqual(newUsers);
        });
    });

    describe('addUser', () => {
        it('should call setState', () => {
            const e = {
                preventDefault: jest.fn(),
                target: [{
                    value: "Kurt",
                }]
            };
            expect(wrapper.state('users')).toEqual(users);
            wrapper.instance().addUser(e);

            expect(addUserSpy).toHaveBeenCalledWith('Kurt');
        });
    });

    describe('onDragEnd', () => {
        it('should reset order when no destination', () => {
            const result = {};

            wrapper.instance().onDragEnd(result);
            let user_1 = wrapper.state('users')[0];
            let user_2 = wrapper.state('users')[1];
            let user_3 = wrapper.state('users')[2];

            expect(user_1.name).toEqual("Bob Ducca");
            expect(user_2.name).toEqual("Dabney Coleperson");
            expect(user_3.name).toEqual("Llarn Bleb-Raag");
        });

        it('should keep same order', () => {
            const result = {
                source: {
                    index: 0,
                },
                destination: {
                    index: 0,
                },
            };

            wrapper.instance().onDragEnd(result);
            let user_1 = wrapper.state('users')[0];
            let user_2 = wrapper.state('users')[1];
            let user_3 = wrapper.state('users')[2];

            expect(user_1.name).toEqual("Bob Ducca");
            expect(user_2.name).toEqual("Dabney Coleperson");
            expect(user_3.name).toEqual("Llarn Bleb-Raag");
        });

        it('should swap order', () => {
            const result = {
                source: {
                    index: 0,
                },
                destination: {
                    index: 1,
                },
            };

            wrapper.instance().onDragEnd(result);
            let user_1 = wrapper.state('users')[0];
            let user_2 = wrapper.state('users')[1];
            let user_3 = wrapper.state('users')[2];

            expect(user_1.name).toEqual("Dabney Coleperson");
            expect(user_2.name).toEqual("Bob Ducca");
            expect(user_3.name).toEqual("Llarn Bleb-Raag");
        });
    });

    describe('getItemStyle', () => {
        it('should return when dragging', () => {
            let draggableStyle = {
                boxSizing: "border-box",
                pointerEvents: "none",
                position: "fixed",
                transition: "opacity 0.2s cubic-bezier(0.2, 0, 0, 1)",
                zIndex: 5000,
            };

            let itemStyle = wrapper.instance().getItemStyle(true, draggableStyle);
            expect(itemStyle.background).toBe("var(--highlight-color)");
            expect(itemStyle).toMatchSnapshot();
        });

        it('should return when not dragging', () => {
            let draggableStyle = {
                boxSizing: "border-box",
                pointerEvents: "none",
                position: "fixed",
                transition: "opacity 0.2s cubic-bezier(0.2, 0, 0, 1)",
                zIndex: 5000,
            };

            let itemStyle = wrapper.instance().getItemStyle(false, draggableStyle);
            expect(itemStyle.background).toBe("transparent");
            expect(itemStyle).toMatchSnapshot();
        });
    });

    describe('getListStyle', () => {
        it('should return when isDraggingOver', () => {
            let listStyle = wrapper.instance().getListStyle(true);

            expect(listStyle.background).toBe("rgba(0,0,0,0.2)");
            expect(listStyle).toMatchSnapshot();
        });

        it('should return when not isDraggingOver', () => {
            let listStyle = wrapper.instance().getListStyle(false);

            expect(listStyle.background).toBe("");
            expect(listStyle).toMatchSnapshot();
        });
    });

    describe('reOrder', () => {
        it('move from 0 to 1', () => {
            let reOrdered = wrapper.instance().reOrder(users, 0, 1);
            expect(reOrdered).toMatchSnapshot();
        });

        it('move from 1 to 2', () => {
            let reOrdered = wrapper.instance().reOrder(users, 1, 2);
            expect(reOrdered).toMatchSnapshot();
        });

        it('move from 2 to 0', () => {
            let reOrdered = wrapper.instance().reOrder(users, 2, 0);
            expect(reOrdered).toMatchSnapshot();
        });
    });

    describe('render', () => {
        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
