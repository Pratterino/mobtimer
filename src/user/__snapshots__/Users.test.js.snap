// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users getItemStyle should return when dragging 1`] = `
Object {
  "background": "var(--highlight-color)",
  "boxSizing": "border-box",
  "pointerEvents": "none",
  "position": "fixed",
  "transition": "opacity 0.2s cubic-bezier(0.2, 0, 0, 1)",
  "userSelect": "none",
  "zIndex": 5000,
}
`;

exports[`Users getItemStyle should return when not dragging 1`] = `
Object {
  "background": "transparent",
  "boxSizing": "border-box",
  "pointerEvents": "none",
  "position": "fixed",
  "transition": "opacity 0.2s cubic-bezier(0.2, 0, 0, 1)",
  "userSelect": "none",
  "zIndex": 5000,
}
`;

exports[`Users getListStyle should return when isDraggingOver 1`] = `
Object {
  "background": "rgba(0,0,0,0.2)",
  "display": "inline-flex",
  "overflow": "auto",
}
`;

exports[`Users getListStyle should return when not isDraggingOver 1`] = `
Object {
  "background": "",
  "display": "inline-flex",
  "overflow": "auto",
}
`;

exports[`Users reOrder move from 0 to 1 1`] = `
Array [
  Object {
    "name": "Dabney Coleperson",
  },
  Object {
    "name": "Bob Ducca",
  },
  Object {
    "name": "Llarn Bleb-Raag",
  },
]
`;

exports[`Users reOrder move from 1 to 2 1`] = `
Array [
  Object {
    "name": "Bob Ducca",
  },
  Object {
    "name": "Llarn Bleb-Raag",
  },
  Object {
    "name": "Dabney Coleperson",
  },
]
`;

exports[`Users reOrder move from 2 to 0 1`] = `
Array [
  Object {
    "name": "Llarn Bleb-Raag",
  },
  Object {
    "name": "Bob Ducca",
  },
  Object {
    "name": "Dabney Coleperson",
  },
]
`;

exports[`Users render should match snapshot when active 1`] = `
<div
  className="Users active"
>
  <DragDropContext
    onDragEnd={[Function]}
  >
    <Connect(Droppable)
      direction="horizontal"
      droppableId="droppable"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={false}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
  </DragDropContext>
  <section
    className="users__footer"
  >
    <FontAwesomeIcon
      border={false}
      className="pointer icon-shadow active"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f050",
            "M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z",
          ],
          "iconName": "fast-forward",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size="2x"
      spin={false}
      swapOpacity={false}
      symbol={false}
      title="Next user!"
      transform={null}
    />
  </section>
</div>
`;

exports[`Users render should match snapshot when inactive 1`] = `
<div
  className="Users"
>
  <DragDropContext
    onDragEnd={[Function]}
  >
    <Connect(Droppable)
      direction="horizontal"
      droppableId="droppable"
      getContainerForClone={[Function]}
      ignoreContainerClipping={false}
      isCombineEnabled={false}
      isDropDisabled={false}
      mode="standard"
      renderClone={null}
      type="DEFAULT"
    >
      <Component />
    </Connect(Droppable)>
  </DragDropContext>
  <section
    className="users__footer"
  >
    <FontAwesomeIcon
      border={false}
      className="pointer icon-shadow inactive"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f050",
            "M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z",
          ],
          "iconName": "fast-forward",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size="2x"
      spin={false}
      swapOpacity={false}
      symbol={false}
      title="Next user!"
      transform={null}
    />
  </section>
</div>
`;
